### PR TITLE
Feat/fetch dr wasm

### LINKED ISSUE #38: ✨ Fetch DataRequest WASM from ID
## ✨ Feature

Fetch Data Request WASM binaries (when required from Data Request ID) from the WASM Storage contract. Cache & Execute them.

## Motivation

This allows user the full flexibility on how the data request should be executed

## Implementation

* When the VM gets triggered with a `WasmId::Id` try to fetch the Wasm file from the main chain if it does not already exist in the wasm_cache. Then if it doesn't exist, store the Wasm binary in cache.
* This should be done inside of WASM itself.
    * Create a new `wasm_store(bytes: u8)` VM import that can be called to store a WASM binary
    * Create a new `wasm_exists(id: Hash)` VM import that can be called to check if a binary exists in the node's storage/cache.
    * Fetching can just be done using the `chain_view` method.

### PR DESCRIPTION
**Pull request by gluax**
## Motivation

To store WASM binaries from the chain in the overlay node local cache.

## Explanation of Changes

I added the new import methods `wasm_exists` and `wasm_store`, which re-used some existing logic.

So that now has been turned into a function. 

I also added a new method, `store_wasm_locally`, in `data_request.rs`  of the consensus binary. However, should this be moved to its own WASM binary for the `Wasm::Id(String)` in `runtime_job.rs` of the node? Or should it be usable from both places?

I didn't add it to the workflow anywhere within `data_request.rs` since we are still figuring out the exact storage.

I also noticed we were overly deriving(mostly serialize or deserialize on objects where we only did one of those) on some objects and putting unnecessary bloat so I removed those.

Lastly,  I regenerated our CLI readme, since when I went to use it I saw it was out of date and got confused.

## Testing

Tested manually by calling the function in the flow temporarily to see if it all works. And ran existing tests.

## Related PRs and Issues

Closes #38.

### CODE REVIEW

**Comment by gluax** [comment]
Rude Rust updated and changed clippy settings the day I opened this PR. I'll fix those new clippy issues >.<.

**Review by Thomasvdam — COMMENTED** [review]
Just some questions/minor remarks. There's a bit too much going on that I don't fully understand yet to feel comfortable approving. :)

File: chains/src/seda_chain.rs [hunk 1]
```diff
@@ -61,12 +61,12 @@ impl SedaChain {
     ///
     /// * `private_key_bytes` - A byte slice that holds the private key of the sender.
     /// * `config`            - A struct that holds the Cosmos chain configuration.
-    pub fn new(config: SedaChainConfig) -> Result<Arc<Self>> {
+    pub fn new(config: SedaChainConfig) -> Result<Rc<Self>> {
```
[the remark this answers is not in the document]
Comment (gluax) [reply chains/src/seda_chain.rs:64]: The latest Rust update stabilized some `clippy`, which affected our CI. The reasoning being `Arc` is not `Send`.

File: wasm/consensus/src/models/data_request.rs [hunk 2]
```diff
@@ -30,9 +31,42 @@ pub trait DataRequestMethods {
...
     fn get_result(&self) -> Result<Option<String>>;
     fn post(&self) -> Result<()>;
+    fn store_wasm_locally(id: &str) -> Result<()>;
 }
 
 impl DataRequestMethods for DataRequest {
+    // Checks if the WASM file exists locally
+    // and if doesn't gets it then stores it.
+    // TODO need to call this but unsure of where in the flow
+    fn store_wasm_locally(id: &str) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (gluax) [reply wasm/consensus/src/models/data_request.rs:41]: Naming is not my strong suit. If you have any ideas let me know :)

File: wasm/consensus/src/models/data_request.rs [hunk 3]
```diff
@@ -30,9 +31,42 @@ pub trait DataRequestMethods {
...
+        if !exists {
+            log!(Level::Debug, "WASM Bin with id `{id}` does not exist storing...");
+
+            wasm_store(&json_value.binary);
+
+            return Ok(());
+        }
+
+        log!(Level::Debug, "WASM Bin with id `{id}` already exists");
+        Ok(())
```
[the remark this answers is not in the document]
Comment (gluax) [reply wasm/consensus/src/models/data_request.rs:67]: The problem is the hash used on the contract is different from the one used locally. So we need to pull the bytes to construct the hash for the wasmer cache.

File: node/src/runtime_job.rs [hunk 4]
```diff
@@ -59,6 +59,8 @@ impl Handler<RuntimeJob> for RuntimeWorker {
                 allowed_configured_imports.unwrap_or(CORE_IMPORTS.to_vec()),
             ),
             // TODO: Placeholder for when we are going to fetch and cache binaries
+            // @Franklin Should I move the method I made in data_request.rs to another wasm binary?
```
Comment (FranklinWaller) [inline node/src/runtime_job.rs:62]: This should load the module from cache

File: runtime/core/src/context.rs [hunk 5]
```diff
@@ -58,6 +58,12 @@ impl VmContext {
         self.memory().view(store)
     }
 
+    /// Provides safe access to the memory
+    /// (it must be initialized before it can be used)
+    pub fn memory_view_mut<'a>(&'a mut self, store: &'a mut impl AsStoreRef) -> MemoryView<'a> {
```
Comment (FranklinWaller) [inline runtime/core/src/context.rs:63]: Delete me

File: runtime/core/src/vm_import_types.rs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 use lazy_static::lazy_static;
 
+// TODO auto generate this list somehow?
```
Comment (FranklinWaller) [inline runtime/core/src/vm_import_types.rs:3]: We shouldn't do this

File: wasm/consensus/src/models/data_request.rs [hunk 7]
```diff
@@ -30,9 +31,42 @@ pub trait DataRequestMethods {
...
 
 impl DataRequestMethods for DataRequest {
+    // Checks if the WASM file exists locally
+    // and if doesn't gets it then stores it.
+    // TODO need to call this but unsure of where in the flow
+    fn store_wasm_locally(id: &str) -> Result<()> {
+        // TODO rn this contract is hard coded but we should fix this. Another config value?
+        // This contract is different from the one in the env variables we currently have.
+        let json_resp = main_chain_view(
+            "seda1ufs3tlq4umljk0qfe8k5ya0x6hpavn897u2cnf9k0en9jr7qarqqdskv5z".to_string(),
```
[the remark this answers is not in the document]
Comment (FranklinWaller) [reply wasm/consensus/src/models/data_request.rs:45]: For now yes, let's put it in a config value. Later on this is all auto fetched from the cosmos sdk anyway

File: wasm/consensus/src/models/data_request.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (FranklinWaller) [reply wasm/consensus/src/models/data_request.rs:67]: Yeah for now let's keep it at this, later on it should then just be a wasm_exists(hash)

**Review by FranklinWaller — APPROVED** [review]

**Comment by gluax** [comment]
The code coverage CI was running indefinitely. Not sure why. We weren't even using it while the repo's were private. We should fix that/remove it in another pr/issue.

**Review by FranklinWaller — APPROVED** [review]

**Pull request merged**